File: tower_of_hanoi.js
"use strict";

const towerOfHanoi = function (height) {
  const moveRing = function (ring, target) {
    const home = rings[ring];
    const spare = [1, 2, 3].find((tower) => {
      return tower !== target && tower !== home;
    });
    if (rings[ring + 1]) {
        if (rings[ring + 1] !== spare) {
          moveRing((ring + 1), spare);
        }
    }
    rings[ring] = target;
    console.log(`Moved ring ${ring} from ${home} to ${target}.`);
    moves++;
    if (rings[ring + 1]) {
      moveRing((ring + 1), target);
    }
  }

  let moves = 0;
  const rings = [];
  for (let i = 0; i < height; i++) {
    rings.push(1);
  };
  console.log(rings);
  moveRing(0, 3);
  console.log(rings);
  console.log(`Moves: ${moves}`);
}

// towerOfHanoi(3);
// towerOfHanoi(4);
// towerOfHanoi(5);
// towerOfHanoi(6);
// towerOfHanoi(7);
towerOfHanoi(8);
// towerOfHanoi(9);
// towerOfHanoi(10);
// towerOfHanoi(11);

File: tower_of_hanoi.cpp
#include <iostream>

int main(int argc, char const *argv[]) {
    using namespace std;
    int heightInput;
    // cout << "Please enter the number of rings in your Tower of Hanoi puzzle." << endl;
    // cin >> heightInput;
   if (argc > 1) {
     heightInput = atoi(argv[1]);
   } else {
     heightInput = 12;
   }
    // cout << heightInput << endl;
    int moves = 0;
    int rings[heightInput + 1];
    for (int i = 0; i < heightInput; i++) {
        rings[i] = 1;
    }
    rings[heightInput] = 0;
    function<void(int,int)> moveRing = [&moves, &rings, &moveRing] (int ringIndex, int target) mutable {
        int home = rings[ringIndex];
        int spare = 0;
        for (int i = 1; i < 4; i++) {
            if (i != target && i != home) {
                spare = i;
            }
        }
        if (rings[ringIndex + 1] != 0) {
            if (rings[ringIndex + 1] != spare) {
                moveRing((ringIndex + 1), spare);
            }
        }
        rings[ringIndex] = target;
        // cout << "Moved ring " << ringIndex << " from " << home << " to " << target << "." << endl;
        moves++;
        if (rings[ringIndex + 1]) {
            moveRing((ringIndex + 1), target);
        }
    };
    // for (int i = 0; i < heightInput; i++) {
    //     cout << rings[i] << " ";
    // }
    // cout << endl;
    moveRing(0, 3);
    // for (int i = 0; i < heightInput; i++) {
    //     cout << rings[i] << " ";
    // }
    // cout << endl;
    // cout << "Moves: " << moves << endl;

    return 0;
}
